### PR TITLE
TUI: add more keycodes (especially keypad keys)

### PR DESCRIPTION
**Pull request by FractalWire**
Follow up on #9780

This PR add the mapping for keymap function on rxvt terminals.

### CODE REVIEW

File: src/nvim/keymap.h [hunk 1]
```diff
@@ -372,7 +372,7 @@ enum key_extra {
 #define K_KPAGEUP       TERMCAP2KEY('K', '3')   /* keypad pageup (upper R.) */
 #define K_KPAGEDOWN     TERMCAP2KEY('K', '5')   /* keypad pagedown (lower R.) */
 
-#define K_KCENTER       TERMCAP2KEY('K', '2')   /* WIP: correct name ? */
+#define K_KORIGIN       TERMCAP2KEY('K', '2')   // RFC: correct name ?
```
Comment (justinmk) [inline src/nvim/keymap.h]: :+1:

**Comment by erw7** [comment]
What is `kUp`, `KDown`, `kLeft`, `KRight`, `kOrigin`? Looking at [termkey.c](https://github.com/neovim/libtermkey/blob/d9928fdb7c079ab429e4132d6c35ca32f893b66f/termkey.c#L56), `termkey_strfkey` never returns them. Therefore, defining them in `keymap.c` has no meaning.

Also, for `KP0`, `KPPlus`, etc., I think that libtermkey should be changed to return `k0`, `kPlus`,etc., considering compatibility with vim.

Also, I wonder if `k7(KP7)`, etc., should be mapped to `kHome`, etc. Certainly it seems strange that different terminal emulators return different key codes for the same key, but this is the result of each terminal emulator emulating a different terminal. I think `k7` should be mapped to` K_K7` like as vim and mapping of this key in neovim should be as follows:

```vim
nnoremap <k7> :echo 'k7'<CR>
```

**Comment by FractalWire** [comment]
> What is kUp, KDown, kLeft, KRight, kOrigin? Looking at termkey.c, termkey_strfkey never returns them. Therefore, defining them in keymap.c has no meaning.

Indeed, it never does, but the same could be said for kHome, kEnd, kPageup, kPlus and similar.
I added those naming for it to be the 'official' name to use when a user want to map a keypad key.

> Also, for KP0, KPPlus, etc., I think that libtermkey should be changed to return k0, kPlus,etc., considering compatibility with vim.

It might, if it does not break compatibility with other programs along the way.
I would rename KP0 to kInsert instead, given that termkey return a function key here, not the 0 value.

> Also, I wonder if k7(KP7), etc., should be mapped to kHome, etc. [...] I think k7 should be mapped to K_K7 like as vim and mapping of this key in neovim should be as follows:

Well, I think it would be quite confusing for the user. In fact, it never occured to me that I should use `<k7>` to access the keypad 'Home' function, especially due to the fact that a `<kHome>` is mentionned in the documentation.
Furthermore, considering that your mapping behaves differently in vim and in gvim, I think we could consider this as a bug in vim, or at least a serious flaw.
Now I don't know exactly how gui integration is done in neovim, but I guess gui applications might want to include both `<k7>` and `<kHome>` mappings.

**Comment by justinmk** [comment]
> Also, for KP0, KPPlus, etc., I think that libtermkey should be changed to return k0, kPlus,etc., considering compatibility with vim.

Can't Nvim map the libtermkey names to some other name? E.g. `KPPlus` resolves to `KPlus`.

> Furthermore, considering that your mapping behaves differently in vim and in gvim, I think we could consider this as a bug in vim, or at least a serious flaw.

Yes, if we can match GUI behavior then that is probably best.

@FractalWire what's remaining here? If you want to add a test, this approach at least tests that the "alias" keycodes map to the expected keycodes:

```diff
diff --git a/test/functional/ui/input_spec.lua b/test/functional/ui/input_spec.lua
index 850efed28291..6b8d52161006 100644
--- a/test/functional/ui/input_spec.lua
+++ b/test/functional/ui/input_spec.lua
@@ -30,6 +30,7 @@ describe('mappings', function()
     add_mapping('<c-s-a-d-up>', '<c-s-a-d-up>')
     add_mapping('<c-d-a>', '<c-d-a>')
     add_mapping('<d-1>', '<d-1>')
+    add_mapping('<kHome>', '<kHome>')
   end)
-
   it('ok', function()
@@ -48,7 +49,10 @@ describe('mappings', function()
     check_mapping('<c-d-a>', '<c-d-a>')
     check_mapping('<d-c-a>', '<c-d-a>')
     check_mapping('<d-1>', '<d-1>')
+    check_mapping('<kHome>', '<kHome>')
+    check_mapping('<kPHome>', '<kHome>')  -- <kPHome> resolves to <kHome>
   end)
+
 end)
-
 describe('feeding large chunks of input with <Paste>', function()
```

You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#running-tests) like this:

    TEST_FILE=test/functional/ui/input_spec.lua make functionaltest

**Comment by FractalWire** [comment]
> @FractalWire what's remaining here?

Well, I was considering wether or not I should match the behaviour of kHome for the keypad arrows.

For the moment, the keypad arrow keys replicate the behaviour of the normal arrow keys because of these [lines](https://github.com/neovim/neovim/pull/9793/commits/d7f785f13116974acf6d150e256771dca7ef87b6#diff-1c25b3c9884177e3e12380a2ced7d51bR1468).

The only drawback that I noticed to this seems that when I press `<ctrl+v><kUp>`, `<Up>` is printed instead of `<kUp>`, which is annoying if a user wants to map these keys.

So, I could do something similar to what is done with kHome, that is put kHome everywhere Home is being used, [see that for example](https://github.com/neovim/neovim/blob/master/src/nvim/edit.c#L998).

The thing is, I'm not sure that this approach is the best and replicating that for the keypad arrow keys would require extensive changes (edit.c, normal.c, ex_getln.c, etc.), and given that I don't understand fully the code, my changes would be prone to errors.

So ideally, and for simplicity sake, we would have a vgetc that behaves like it behaves with my changes except when it is invoked by ctrl-v. And so far, I don't know how to do that. 
Doing that would also mean that having K_KHOME (or any keypad mapping) whenever K_HOME is used in various parts of the code would probably not be necessary anymore.

Also, I wonder if I should add a mapping for `<C-kUp>` and similar and how to do that.

What do you think ?

**Comment by justinmk** [comment]
> For the moment, the keypad arrow keys replicate the behaviour of the normal arrow keys because of these [lines](https://github.com/neovim/neovim/pull/9793/commits/d7f785f13116974acf6d150e256771dca7ef87b6#diff-1c25b3c9884177e3e12380a2ced7d51bR1468).
> 
> The only drawback that I noticed to this seems that when I press `<ctrl+v><kUp>`, `<Up>` is printed instead of `<kUp>`, which is annoying if a user wants to map these keys.

Well that makes sense, because `vgetc` maps kUp to Up. Thus these are considered identical keycodes from Nvim's perspective and for mapping purposes (`:nnoremap`).
If you want them to be separate keycodes then `vgetc` should _not_ map `K_KUP` to `K_UP`.
 
> So, I could do something similar to what is done with kHome, that is put kHome everywhere Home is being used, [see that for example](https://github.com/neovim/neovim/blob/master/src/nvim/edit.c#L998).
> 
> The thing is, I'm not sure that this approach is the best and replicating that for the keypad arrow keys would require extensive changes (edit.c, normal.c, ex_getln.c, etc.), and given that I don't understand fully the code, my changes would be prone to errors.

There's not really any other way. If you want to introduce a new keycode (K_KUP, etc.) then the various modes must be updated to have "default" behavior in the absence of a mapping.

I think this is worth doing, but I would save it for a followup pull request. So keep this one focused as it is currently, it fixes a problem without any drawbacks. Then we can merge this quickly and focus on the next PR.

- For this PR, I recommend not aliasing  in `vgetc` any keycodes that you want to be separately mappable and visible with ctrl-v. 
- So users can map those keycodes, but otherwise those keycodes will be ignored by default. 
- In the next PR, we can work on adding the "default" behavior in edit.c etc.

> So ideally, and for simplicity sake, we would have a vgetc that behaves like it behaves with my changes except when it is invoked by ctrl-v. And so far, I don't know how to do that.
> Doing that would also mean that having K_KHOME (or any keypad mapping) whenever K_HOME is used in various parts of the code would probably not be necessary anymore.

That wouldn't really make sense. ctrl-v shows whatever actually is passed through the rest of the application. Showing something else would be misleading.

> Also, I wonder if I should add a mapping for `<C-kUp>` and similar and how to do that.

That should already happen automatically, assuming your terminal (rxvt or whatever) actually sends bytes for that chord. You can test that Nvim does the right thing by using the `nvim_input` function:

    :nnoremap <c-kup> :echo 'got c-kup'<cr>
    :call nvim_input('<c-kup>')

> On a side note, if I run `make functionaltest`, I've got errors that seems unrelated to my changes.

Need to see the errors.

**Comment by FractalWire** [comment]
> Well that makes sense, because vgetc maps kUp to Up. Thus these are considered identical keycodes from Nvim's perspective and for mapping purposes (:nnoremap).

Well, that is weird because I can do : 
```
nnoremap <Up> :echo 'Up'<CR>
nnoremap <kUp> :echo 'kUp'<CR>
```
And I have both mapping firing properly. Same thing for the math symbols previously in the switch.
Ctrl-v on the other hand, returns `<Up>` for both keys...

> assuming your terminal (rxvt or whatever) actually sends bytes for that chord.
Ah, yes, that is why it's not working here...

> I think this is worth doing, but I would save it for a followup pull request.
:+1:

**Comment by justinmk** [comment]
> Well, that is weird because I can do :
> 
> ```
> nnoremap <Up> :echo 'Up'<CR>
> nnoremap <kUp> :echo 'kUp'<CR>
> ```
> And I have both mapping firing properly. Same thing for the math symbols previously in the switch.
> Ctrl-v on the other hand, returns `<Up>` for both keys...

My mistake. This comment in `vgetc` explains it:

```c
      /* a keypad or special function key was not mapped, use it like
       * its ASCII equivalent */
```

But that also explains why that code path applies to ctrl-v, when the keycode isn't mapped.

**Review by justinmk — APPROVED** [review]

File: test/symbolic/klee/nvim/keymap.c [hunk 2]
```diff
@@ -236,13 +247,22 @@ static const struct key_name_entry {
   { K_PAGEUP,          "PageUp" },
   { K_PAGEDOWN,        "PageDown" },
   { K_KPAGEUP,         "kPageUp" },
+  { K_KPAGEUP,         "KP9" },
   { K_KPAGEDOWN,       "kPageDown" },
+  { K_KPAGEDOWN,       "KP3" },
+  { K_KORIGIN,         "kOrigin" },  // RFC: Correct name ?
```
Comment (justinmk) [inline test/symbolic/klee/nvim/keymap.c]: I don't think we need "RFC" here anymore.

Comment (FractalWire) [reply test/symbolic/klee/nvim/keymap.c]: Ah, yes I forgot that one

**Comment by justinmk** [comment]
Merged, thank you @FractalWire . 

I got a keyboard with a keypad on it,  this  works nicely! Some things I noticed (which aren't regressions, since keypad doesn't work at all before this PR):

-  keypad `+` doesn't work (sends `^[Ol`). 
    - TODO: We should at least _log_ ignored things...
- in `:terminal cat -v`, keys 2/4/5/6/8 send escape sequences, whereas all other keys (`0 1 3 7 9 + <Enter>. /*-`) send the respective ASCII byte.
    - Maybe `terminal.c` needs a change.

**Comment by FractalWire** [comment]
> Merged, thank you @FractalWire .

My pleasure ! I'm pretty glad to have been able to contribute a few bytes to this project. I kinda want to brag about this new keypad feature now, haha !

> keypad + doesn't work (sends ^[Ol).
TODO: We should at least log ignored things...

Probably something to do with your terminfo, or the fact that your '+' is really recognize as a ',', according to this : https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h2-VT220-Style-Function-Keys
I did not add 'Comma' or 'Equal' to the new keycodes, but that might be something to do, considering they exist [in termkey](https://github.com/neovim/libtermkey/blob/9b810707e1455b6f3a85204cbd6d677001443da7/termkey.c#L99)

> in :terminal cat -v, keys 2 4 5 6 8 send escape sequences, whereas all other keys (0 1 3 7 9 + <Enter>. /*-) send the respective ASCII byte.
Maybe terminal.c needs a change.

Ah, yes, I revert those changes, considering they might be better put in the next PR for the arrow key mapping in the various modes.

Here is a [small commit](https://github.com/FractalWire/neovim/commit/dcf6f65b896d8504a93edb34a7356a5fd88a106d) that changes the behaviour of the keypad keys to match their 'normal' function key counterpart.
Note that it is not the proper way of doing this, because it now means that we 'hide' the real key being pressed to the application running in the `:term` instance.
To fix that I'd need to understand a bit better where these `VTERM_KEY_*` come from. I tried to look at libtermkey but that does not seem to be there either.

Should I do a new PR for these changes and then another one for the keypad arrow support in the other modes ?

**Comment by justinmk** [comment]
> To fix that I'd need to understand a bit better where these `VTERM_KEY_*` come from. I tried to look at libtermkey but that does not seem to be there either.

They come from libvterm . https://github.com/neovim/libvterm/blob/master/include/vterm_keycodes.h

> Should I do a new PR for these changes and then another one for the keypad arrow support in the other modes ?

That will make things go faster. Also helps for bisecting, in case anything goes wrong later.

**Pull request closed without merging**